core/math/color.cpp

```diff
@@ -109,35 +109,29 @@ uint64_t Color::to_rgba64() const {
 	return c;
 }
 
-String _to_hex(float p_val) {
+void _append_hex(float p_val, String &string) {
 	int v = Math::round(p_val * 255.0f);
 	v = CLAMP(v, 0, 255);
-	String ret;
 
 	for (int i = 0; i < 2; i++) {
-		char32_t c[2] = { 0, 0 };
 		int lv = v & 0xF;
 		if (lv < 10) {
-			c[0] = '0' + lv;
+			string += '0' + lv;
 		} else {
-			c[0] = 'a' + lv - 10;
+			string += 'a' + lv - 10;
```

Comment:
The size is known, so we can get rid of `+=` and preallocate string:

```c++
void _append_hex(float p_val, char32_t *string) {
	int v = Math::round(p_val * 255.0f);
	v = CLAMP(v, 0, 255);

	for (int i = 0; i < 2; i++) {
		int lv = v & 0xF;
		if (lv < 10) {
			string[i] = '0' + lv;
		} else {
			string[i] = 'a' + lv - 10;
		}

		v >>= 4;
	}
}

String Color::to_html(bool p_alpha) const {
	String txt;
	txt.resize(p_alpha ? 9 : 7);
	char32_t *ptr = txt.ptrw();
	_append_hex(r, ptr);
	_append_hex(g, ptr + 2);
	_append_hex(b, ptr + 4);
	if (p_alpha) {
		_append_hex(a, ptr + 6);
	}
	ptr[p_alpha ? 8 : 6] = 0x00;
	return txt;
}
```

Master: 37349 ms
This PR: 9711 ms
Prealloc: 1963 ms

Reply:
Makes me wonder if this can even get some more speed now : 

```
char hexa[] = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'};
void _append_hex(float p_val, char32_t *string) {
	int v = Math::round(p_val * 255.0f);
	v = CLAMP(v, 0, 255);

	string[i] = hexa[v & 0xF];
	string[i+1] = hexa[(v >> 4) & 0xF];
}
```

Reply:
> Makes me wonder if this can even get some more speed now 

A bit, `1790 ms` with the same config on the same system.

Reply:
(thank you for running it)